### PR TITLE
HDDS-8939. [Snapshot] isBlockLocationSame check should be skipped if object is not OmKeyInfo.

### PR DESCRIPTION
**Pull request by swamirishi**
## What changes were proposed in this pull request?
As part of [HDDS-8841](https://issues.apache.org/jira/browse/HDDS-8841), we added a check if block location is same for key object. Somehow isBlockLocationSame function is getting called for non-OmKeyInfo object. We should skip this check is object are not OmKeyInfo. In the case it is a directory we should check if the directory acls are the same otherwise we should return a modify entry for the directory.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8939

## How was this patch tested?
Unit Tests.

### CODE REVIEW

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java [hunk 1]
```diff
@@ -1145,20 +1148,22 @@ void addToObjectIdMap(Table<String, ? extends WithParentObjectId> fsTable,
   }
 
   private SnapshotDiffObject createDiffObjectWithOldName(
-      long objectId, String oldName, SnapshotDiffObject diffObject) {
+      long objectId, String oldName, SnapshotDiffObject diffObject,
+      boolean isDirectory) {
     SnapshotDiffObjectBuilder builder = new SnapshotDiffObjectBuilder(objectId)
-        .withOldKeyName(oldName);
+        .withOldKeyName(oldName).isDirectory(isDirectory);
```
Comment (prashantpogde) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:1154]: .setDirectory() ??

Comment (swamirishi) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java]: done. Changed the name

**Review by prashantpogde — APPROVED** [review]
thank you @swamirishi for making the changes. LGTM

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java [hunk 2]
```diff
@@ -1439,6 +1448,34 @@ long generateDiffReport(
...
+    Objects.requireNonNull(toObjectName, "toObjectName is null.");
+
+    final WithObjectID fromObject = fromSnapshotTable.get(fromObjectName);
+    final WithObjectID toObject = toSnapshotTable.get(toObjectName);
+    if ((fromObject instanceof OmKeyInfo) && (toObject instanceof OmKeyInfo)) {
+      return !SnapshotDeletingService.isBlockLocationInfoSame(
+          (OmKeyInfo) fromObject, (OmKeyInfo) toObject);
+    } else if ((fromObject instanceof OmDirectoryInfo)
+        && (toObject instanceof OmDirectoryInfo)) {
+      return !areAclsSame((OmDirectoryInfo) fromObject,
```
Comment (DaveTeng0) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:1465]: Is renaming/moving a directory considered as 'modified'?
For example:
1. Rename /path_1/dir_1 to /path_1/dir_2
2. Move /path_1/dir_1 to /path_2/dir_1

Comment (swamirishi) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:1465]: Rename will not modify acls. Renames will occur as RENAME entry in Snapdiff. You can take a look at the test case.

**Pull request merged**